src/transformStateWithClones.js

```diff
@@ -6,8 +6,51 @@
  *
  * @return {Object[]}
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+function transformStateWithClones(state = {}, actions = []) {
+  const clonedState = JSON.parse(JSON.stringify(state));
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(clonedState, action.extraData, stateHistory);
+        break;
+      case 'removeProperties':
+        removeProperties(clonedState, action.keysToRemove, stateHistory);
+        break;
+      case 'clear':
+        clearProperties(clonedState, stateHistory);
```

Comment:
Instead of calling separate functions for each action type and passing the clonedState and stateHistory as arguments, consider performing the action directly within the case block to avoid unnecessary function calls and to make the code more readable. Additionally, this will help in ensuring that the cloned state is not mutated across different actions.